### ISSUE TITLE
ERROR: [udp] sendto_remote: Operation not permitted

### ISSUE THREAD

**Issue by cokebar** [shadowsocks/shadowsocks-libev#1619]
### What version of shadowsocks-libev are you using?
3.0.8 （both server side and client side）

### What operating system are you using?
server: Ubuntu 16.04 LTS
client: LEDE 17.01

### What did you do?
Using a LEDE router to proxy both TCP and UDP traffic.
server: `ss-server -c /etc/shadowsocks-libev/config.json -u --fast-open -v`
client: Using @aa65535 's openwrt-shadowsocks & luci-app-shadowsocks. Enable both TCP & UDP proxy.

### What did you expect to see?
No ERROR give out.

### What did you see instead?
An Error shows frequently:
```
ERROR: [udp] sendto_remote: Operation not permitted
```

### What is your config in detail (with all sensitive info masked)?
```
{
    "server":"x.x.x.x",
    "server_port:xxx,
    "password":"xxxxxxx",
    "method":"chacha20-ietf-poly1305",
    "fast_open":true,
    "mtu":1492
}
```

**Comment by madeye** [comment]
TPROXY requires ROOT permission.

**Comment by cokebar** [comment]
@madeye 
1. This error message appears on server side. Even I run ss-server using root,  it still appears. On LEDE router, ss-redir also run as root, and proxy all traffic to ss-server for a WIN10 PC.

Some detailed log (server side ss-server 3.0.8; client side ss-redir 3.0.8):
```
2017-08-08 22:04:10 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:33771
 2017-08-08 22:04:10 INFO: [udp] server receive a packet
 2017-08-08 22:04:10 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:10 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:10 INFO: [udp] cache hit: 74.125.204.101:443 <-> 61.144.101.27:47498
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote_recv_sendto fragmentation
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote_recv_sendto fragmentation
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:33771
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:33771
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:33771
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:33771
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:46252
 2017-08-08 22:04:11 INFO: [udp] server_recv_recvfrom fragmentation
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:33771
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:33771
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:46252
 2017-08-08 22:04:11 INFO: [udp] server_recv_recvfrom fragmentation
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:33771
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] remote receive a packet
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:33771
 2017-08-08 22:04:11 INFO: [udp] server receive a packet
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:11 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:11 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:33771
 2017-08-08 22:04:12 INFO: [udp] server receive a packet
 2017-08-08 22:04:12 INFO: [udp] cache miss: 59.14.108.205:40901 <-> 61.144.101.27:35781
 2017-08-08 22:04:12 ERROR: [udp] sendto_remote: Operation not permitted
 2017-08-08 22:04:13 INFO: [udp] server receive a packet
 2017-08-08 22:04:13 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:13 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:13 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:46252
 2017-08-08 22:04:13 INFO: [udp] remote receive a packet
 2017-08-08 22:04:13 INFO: [udp] remote receive a packet
 2017-08-08 22:04:13 INFO: [udp] remote receive a packet
 2017-08-08 22:04:13 INFO: [udp] server receive a packet
 2017-08-08 22:04:13 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:13 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:13 INFO: [udp] cache hit: 172.217.24.206:443 <-> 61.144.101.27:46252
 2017-08-08 22:04:15 INFO: [udp] server receive a packet
 2017-08-08 22:04:15 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:04:15 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:04:15 INFO: [udp] cache hit: 203.208.51.95:443 <-> 61.144.101.27:46251
 2017-08-08 22:04:15 INFO: [udp] remote receive a packet
 2017-08-08 22:04:16 INFO: accept a connection
 2017-08-08 22:04:16 INFO: connect to gfwsl.geforce.com:80
 2017-08-08 22:04:16 INFO: IPv6 resolv: valid domain but no data of requested type
 2017-08-08 22:04:16 INFO: successfully resolved gfwsl.geforce.com
 2017-08-08 22:04:16 INFO: remote connected
 2017-08-08 22:04:16 INFO: server_recv close the connection
```
2. Version v3.0.6 do not show this error message, but it shows many udp fragmentation.
(server side: ss-server 3.0.6; client side: ss-redir 3.0.8)
```
2017-08-08 22:10:33 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:33 INFO: [udp] server receive a packet
 2017-08-08 22:10:33 INFO: [udp] cache miss: 172.217.24.196:443 <-> 61.144.101.27:55276
 2017-08-08 22:10:33 INFO: accept a connection
 2017-08-08 22:10:33 INFO: connect to 64.233.189.199:80
 2017-08-08 22:10:33 INFO: accept a connection
 2017-08-08 22:10:33 INFO: connect to 64.233.189.199:80
 2017-08-08 22:10:33 INFO: remote connected
 2017-08-08 22:10:33 INFO: remote connected
 2017-08-08 22:10:33 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:34 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:34 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:34 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:34 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:34 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:34 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:34 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:35 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:35 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:35 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:35 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:35 INFO: [udp] server receive a packet
 2017-08-08 22:10:35 INFO: sockaddr_cmp: sin_family equal? 1
 2017-08-08 22:10:35 INFO: sockaddr_cmp: sin_port equal? 1
 2017-08-08 22:10:35 INFO: [udp] cache hit: 64.233.189.199:443 <-> 61.144.101.27:51510
 2017-08-08 22:10:35 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:35 INFO: accept a connection
 2017-08-08 22:10:35 INFO: connect to 8.8.8.8:53
 2017-08-08 22:10:35 INFO: remote connected
 2017-08-08 22:10:36 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:36 INFO: server_recv close the connection
 2017-08-08 22:10:36 INFO: current remote connection: 9
 2017-08-08 22:10:36 INFO: current server connection: 9
 2017-08-08 22:10:36 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:36 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:36 ERROR: [udp] server_recv_recvfrom fragmentation: File exists
 2017-08-08 22:10:36 INFO: accept a connection
```

And When I using PS4's network testing, PS4 also shows there're package framentation problem. 

But when I changed ss-server to version 3.0.8, PS4 won't show package framentation problem. And ss-server's log also do not have this kind of error message.

**Comment by madeye** [comment]
It looks an issue of your system. Try adding these rules to your iptables:

```
iptables -I OUTPUT -t raw -p udp -j NOTRACK
iptables -I PREROUTING -t raw -p udp -j NOTRACK
```

**Comment by madeye** [comment]
Closed due to inactivity.

**Comment by 3xpert** [comment]
operation system is cento 6
After added the about rules of iptables that @madeye posted, I still get the same issues.
2018-02-26 11:15:59 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:00 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:00 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:01 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:02 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:03 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:10 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:10 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:11 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:11 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:13 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:14 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:29 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:30 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:30 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:30 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:32 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:16:33 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 11:18:19 ERROR: remote recv: Connection reset by peer
 2018-02-26 11:33:07 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 11:33:07 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 11:33:07 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 11:33:07 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 11:33:07 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 12:34:05 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 12:52:41 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 12:52:41 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 12:52:41 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 13:47:53 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 13:48:00 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 14:02:59 ERROR: [udp] sendto_remote: Operation not permitted
 2018-02-26 14:06:08 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 14:13:35 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 14:13:35 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 14:13:35 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 14:13:35 [simple-obfs] ERROR: server recv: Connection reset by peer
 2018-02-26 14:13:35 [simple-obfs] ERROR: server recv: Connection reset by peer

**Comment by mghadam** [comment]
I had the same issue with ConfigServer firewall (csf) :

`[udp] sendto_remote: Operation not permitted`

I fixed it by adding the following rule : 

`/usr/sbin/iptables -t filter -I OUTPUT -p udp -j ACCEPT`

It is a generic rule though allowing all UDP outgoing packets in the filter table, one can change ACCEPT to LOG to see the log of dropped UDP packets and add a more specific rule instead.

**Comment by werdender** [comment]
I faced the same issue when had a few IP addresses on a server with shadowsocks server and the 0.0.0.0 IP address in the server config. Client was connecting to an IP address A.A.A.A, but shadowsocks server was trying to send outgoing UDP packets from an IP address B.B.B.B ConfigServer Firewall (CSF) was blocking those packets.

I have fixed it with change 0.0.0.0 to certain IP address (A.A.A.A in the case above).
